Refactor: Move compatibility tooling from container-runtime to runtime-utils and introduce better names

TommyBrosman opened the pull request:
## Description

As part of the work needed to plumb minVersionForCollab down to the DDS layer, this change exposes compatibility tooling (formerly compatUtils.ts) to more layers.

compatUtils has been split into runtime-utils/src/compatibilityBase.ts and container-runtime/src/containerCompatibility.ts. Details:

- compatibilityBase.ts contains the shared parts of the compatibility tooling that can be safely used in all layers.
- containerCompatibility.ts handles Container-specific versioning.

## Breaking Changes

- `MinimumVersionForCollab` has been moved, but the original package (container-runtime) still re-exports it. The re-exported type is deprecated and will be removed in a future release.

## Reviewer Guidance

- I'm looking for feedback around reducing the number of types that need to be moved.
  - This was resolved: compatUtils.ts has been split into layers, which drastically reduced the number of types that needed to be moved.
- Compatibility concerns: this change should be largely backwards-compatible minus the deprecations. Leave a comment if you find changes that could be truly breaking.

TommyBrosman committed:
Moved most of the definitions used by compatUtils.ts from container-runtime to runtime-definitions.

TommyBrosman committed:
Merge branch 'main' into move-compat-utils

Abe27342 commented on packages/runtime/container-runtime/src/index.ts:
```diff
@@ -18,8 +13,27 @@ export {
-export { CompressionAlgorithms, disabledCompressionConfig } from "./compressionDefinitions.js";
+export {
+	/**
+	 * @deprecated Import from `@fluidframework/runtime-definitions/internal` instead.
```
if you want to re-export to reduce churn in PR that's fine though

TommyBrosman replied:
(Also this makes me think I should split this up into multiple exports by API level.)

TommyBrosman replied:
I split up the exports and added deprecation comments to all of the non-internal entries.

TommyBrosman committed:
Deprecations.

TommyBrosman committed:
Added missing test scripts.

TommyBrosman committed:
pnpm-lock update.

TommyBrosman committed:
Changed the type of IGCRuntimeOptions.enableGCSweep.

anthony-murphy commented on packages/runtime/container-runtime/api-report/container-runtime.legacy.alpha.api.md:
```diff
@@ -32,28 +28,12 @@ export enum ContainerMessageType {
-    // (undocumented)
-    readonly summaryOptions: ISummaryRuntimeOptions;
-}
+export { ContainerRuntimeOptions }
```
doing that, i think you can make most of this an additive change, as you'll really only need to move MinimumVersionForCollab

TommyBrosman replied:
(FYI for other reviewers) Tony and I talked, I'm going to try to split compatUtils into "general" and layer-specific sections. That should reduce the code churn.

anthony-murphy commented on packages/runtime/runtime-definitions/src/compatUtils.ts:
```diff
@@ -4,15 +4,15 @@
  */
 
 import { assert } from "@fluidframework/core-utils/internal";
-import { FlushMode } from "@fluidframework/runtime-definitions/internal";
```
we don't usually put code in our definition packages, they are mainly mean to be just interfaces. there are utils packages where code lives. i also don't think most of this should move, there may be some common things are that reusable that could move, but in generally out compat enforcement will need to be layered just like everything else in fluid is layer, so the container runtime specific stuff should stay where it is

TommyBrosman committed:
Split compatUtils into layers and reverted most of the type/declaration moves.

TommyBrosman committed:
Cleanup.

TommyBrosman committed:
More cleanup.

anthony-murphy commented:
> As I was saying offline, it would be nice if the logic to generate configs based on the min collab version is in one place so it doesn't diverge and it's easier to understand and maintain.
I have a hard time seeing how this is possible. Due to the componentized nature of Datastores. Each instance of a dds is solely controlled by a specific datastore and each datastores usage of a dds could want or need different configurations. There is also no way for a datastore author to specify container level configurations, and even if there was, it would be difficult to orchestrate given that datastore can both nested and delay loaded. My thought was we were always going to need a parallel implementation at the datastore level for config and compat. Maybe there is a way to project this upward to the container runtime. it seems plausible to reuse the document schema infrastructure, but since evaluation will always be delayed for datastores it would be pretty complicated to also share the evaluation piece.

TommyBrosman committed:
Renamed files.

anthony-murphy commented on packages/runtime/runtime-definitions/package.json:
```diff
@@ -73,23 +74,32 @@
 		"format:biome": "biome check . --write",
 		"lint": "fluid-build . --task lint",
 		"lint:fix": "fluid-build . --task eslint:fix --task format",
+		"test": "npm run test:mocha",
+		"test:benchmark:report": "mocha --timeout 10s --perfMode --parentProcess --fgrep @Benchmark --fgrep @ExecutionTime --reporter @fluid-tools/benchmark/dist/MochaReporter.js \"./dist/**/*.perf.spec.*js\"",
+		"test:coverage": "c8 npm test",
+		"test:mocha": "npm run test:mocha:esm && echo skipping cjs to avoid overhead - npm run test:mocha:cjs",
+		"test:mocha:cjs": "mocha --recursive \"dist/test/**/*.spec.*js\"",
+		"test:mocha:esm": "mocha --recursive \"lib/test/**/*.spec.*js\"",
+		"test:mocha:verbose": "cross-env FLUID_TEST_VERBOSE=1 npm run test:mocha",
```
i think the code should go in runtime-utils, as we don't normally put code in our definitions packages

agarwal-navin commented on packages/runtime/runtime-definitions/src/compatibilityBase.ts:
```diff
@@ -84,151 +84,17 @@ export type ConfigMap<T extends Record<string, unknown>> = {
 /**
  * Returns a default configuration given minVersionForCollab and configuration version map.
+ *
+ * @internal
```
These functions are utilities not defintions. I believe runtime-utils is the right place for them.

TommyBrosman committed:
Reverted name change (test file: compatUtils.ts).

anthony-murphy commented on packages/runtime/container-runtime/src/index.ts:
```diff
@@ -20,6 +20,13 @@ export {
 export { CompressionAlgorithms, disabledCompressionConfig } from "./compressionDefinitions.js";
+export {
+	/**
+	 * @deprecated Import from `@fluidframework/runtime-definitions/internal` instead.
```
i'm pretty sure this doesn't work, as in the jsdoc tag is completely ignored, so doesn't show up in api report, IntelliSense, and won't trigger es lint rules. i'm not sure if there is a better pattern that actually works for this kind of thing. maybe @Josmithr  knows

TommyBrosman commented on packages/test/test-version-utils/src/describeCompat.ts:
```diff
@@ -31,7 +31,7 @@ import {
 	getCompatVersionedTestObjectProviderFromApis,
 	getDriverInformationWhenNoProviderIsAvailable,
-} from "./compatUtils.js";
+} from "./compatibilityBase.js";
```
Tony and I talked, this file isn't the same "compatUtils" as the minVersionForCollab tooling. It consumes a type from that module, but is otherwise not the same. I'm reverting the rename.

TommyBrosman committed:
- Moved compatibility tooling from runtime-definitions to runtime-utils.
- Deleted a stray file (idCompressorDefinitions.ts) from a previous change.
- Added the packageVersion generation step to runtime-utils.

TommyBrosman commented on packages/runtime/container-runtime/src/index.ts:
```diff
@@ -20,6 +20,16 @@ export {
+	/**
+	 * @deprecated Import from `@fluidframework/runtime-utils/internal` instead.
+	 */
+	MinimumVersionForCollab,
```
I'm new to this process, here's the issue I created: #25085. Let me know if I should tweak anything.

alexvy86 replied:
I'd say just use the scoped package names for clarity, and fix the link to the PR.

TommyBrosman replied:
Done. Thanks for catch the PR link, I'm so used to ADO I often forget the syntax is slightly different in GitHub.

TommyBrosman committed:
- Removed internal type from re-exports.
- Added a changeset describing the deprecation.

TommyBrosman committed:
Removed extraneous file from runtime-definitions and updated the changeset.

scottn12 commented on .changeset/proud-ghosts-doubt.md:
```diff
@@ -0,0 +1,8 @@
+"@fluidframework/runtime-utils": minor
+"__section": deprecation
+---
+Moved MinimumVersionForCollab to runtime-utils
+
+MinimumVersionForCollab has been moved from container-runtime to runtime-utils.
```
nit - full package name
```suggestion
Moved MinimumVersionForCollab to @fluidframework/runtime-utils

MinimumVersionForCollab has been moved from @fluidframework/container-runtime to @fluidframework/runtime-utils.
```

alexvy86 replied:
In addition to Scott's suggestion:

"`MinimumVersionForCollab` has been moved from `@fluidframework/container-runtime` to `@fluidframework/runtime-utils`. The export in `@fluidframework/container-runtime` is now deprecated and will be removed in a future version. Consumers should import it from `@fluidframework/runtime-utils` going forward."

TommyBrosman replied:
That seems more clear. I'll update it. I wasn't sure how specific I needed to be within the context of this repo.

alexvy86 commented on packages/runtime/runtime-utils/src/test/compatUtils.spec.ts:
```diff
@@ -14,9 +14,9 @@ import {
 	type ConfigValidationMap,
 	configValueToMinVersionForCollab,
-} from "../compatUtils.js";
+} from "@fluidframework/runtime-utils/internal";
```
This should just be a path import now, right? Also rename the file to match the new source file.

TommyBrosman replied:
Oops, I missed that when I moved compatUtils back. Will update.

alexvy86 commented on packages/runtime/container-runtime/src/containerCompatibility.ts:
```diff
@@ -0,0 +1,176 @@
+
+/**
+ * Subset of the {@link ContainerRuntimeOptionsInternal} properties which
+ * affect {@link @fluidframework/container-runtime#IDocumentSchemaFeatures}.
```
```suggestion
 * affect {@link IDocumentSchemaFeatures}.
```

alexvy86 commented on packages/runtime/container-runtime/src/containerCompatibility.ts:
```diff
@@ -0,0 +1,176 @@
+/**
+ * Returns the default RuntimeOptionsAffectingDocSchema configuration for a given minVersionForCollab.
+ *
+ * @internal
```
Seems like it's only used within the package so it doesn't need to be tagged `@internal`?

alexvy86 commented on packages/runtime/container-runtime/src/index.ts:
```diff
@@ -20,6 +20,12 @@ export {
 export { CompressionAlgorithms, disabledCompressionConfig } from "./compressionDefinitions.js";
+export {
+	/**
+	 * @deprecated Import from `@fluidframework/runtime-utils/internal` instead.
```
alpha+legacy API, right?
```suggestion
	 * @deprecated Import from `@fluidframework/runtime-utils/legacy` instead.
```

alexvy86 commented on packages/runtime/runtime-utils/package.json:
```diff
@@ -123,7 +124,8 @@
 		"@fluidframework/runtime-definitions": "workspace:~",
-		"@fluidframework/telemetry-utils": "workspace:~"
+		"@fluidframework/telemetry-utils": "workspace:~",
+		"semver-ts": "^1.0.3"
```
Did we check if container-runtime still needs a dep on semver-ts after this move?

TommyBrosman replied:
Yeah, it imports `gt` and `lt` in various modules not related to the API versioning feature.

TommyBrosman committed:
Clarified changeset.

TommyBrosman committed:
Update packages/runtime/container-runtime/src/containerCompatibility.ts

Co-authored-by: Alex Villarreal <716334+alexvy86@users.noreply.github.com>

TommyBrosman committed:
PR feedback.

TommyBrosman committed:
Fixed doc comment with mangled file reference.

alexvy86 reviewed (approved):
Approving for docs. The rest looks ok to me as well but I suggest getting at least one other approval for it.

TommyBrosman committed:
Merge branch 'main' into move-compat-utils

scottn12 reviewed (approved):
LGTM!
